module/zfs/mmp.c

```diff
@@ -205,80 +205,51 @@ typedef enum mmp_vdev_state_flag {
 	MMP_FAIL_WRITE_PENDING	= (1 << 1),
 } mmp_vdev_state_flag_t;
 
-static vdev_t *
-mmp_random_leaf_impl(vdev_t *vd, int *fail_mask)
-{
-	int child_idx;
-
-	if (vd->vdev_ops->vdev_op_leaf) {
-		vdev_t *ret;
-
-		if (!vdev_writeable(vd)) {
-			*fail_mask |= MMP_FAIL_NOT_WRITABLE;
-			ret = NULL;
-		} else if (vd->vdev_mmp_pending != 0) {
-			*fail_mask |= MMP_FAIL_WRITE_PENDING;
-			ret = NULL;
-		} else {
-			ret = vd;
-		}
-
-		return (ret);
-	}
-
-	if (vd->vdev_children == 0)
-		return (NULL);
-
-	child_idx = spa_get_random(vd->vdev_children);
-	for (int offset = vd->vdev_children; offset > 0; offset--) {
-		vdev_t *leaf;
-		vdev_t *child = vd->vdev_child[(child_idx + offset) %
-		    vd->vdev_children];
-
-		leaf = mmp_random_leaf_impl(child, fail_mask);
-		if (leaf)
-			return (leaf);
-	}
-
-	return (NULL);
-}
-
 /*
  * Find a leaf vdev to write an MMP block to.  It must not have an outstanding
  * mmp write (if so a new write will also likely block).  If there is no usable
- * leaf in the tree rooted at in_vd, a nonzero error value is returned, and
- * *out_vd is unchanged.
- *
- * The error value returned is a bit field.
- *
- * MMP_FAIL_WRITE_PENDING
- * If set, one or more leaf vdevs are writeable, but have an MMP write which has
- * not yet completed.
- *
- * MMP_FAIL_NOT_WRITABLE
- * If set, one or more vdevs are not writeable.  The children of those vdevs
- * were not examined.
+ * leaf, a nonzero error value is returned. The error value returned is a bit
+ * field.
  *
- * Assuming in_vd points to a tree, a random subtree will be chosen to start.
- * That subtree, and successive ones, will be walked until a usable leaf has
- * been found, or all subtrees have been examined (except that the children of
- * un-writeable vdevs are not examined).
- *
- * If the leaf vdevs in the tree are healthy, the distribution of returned leaf
- * vdevs will be even.  If there are unhealthy leaves, the following leaves
- * (child_index % index_children) will be chosen more often.
+ * MMP_FAIL_WRITE_PENDING   One or more leaf vdevs are writeable, but have an
+ *                          outstanding MMP write.
+ * MMP_FAIL_NOT_WRITABLE    One or more leaf vdevs are not writeable.
  */
 
 static int
-mmp_random_leaf(vdev_t *in_vd, vdev_t **out_vd)
+mmp_next_leaf(spa_t *spa, vdev_t **leaf_to_use)
 {
-	int error_mask = 0;
-	vdev_t *vd = mmp_random_leaf_impl(in_vd, &error_mask);
+	vdev_t *leaf;
+	vdev_t *starting_leaf;
+	boolean_t found = B_FALSE;
+	int fail_mask = 0;
+
+	ASSERT(list_link_active(&spa->spa_leaf_list.list_head) == B_TRUE);
+	ASSERT(!list_is_empty(&spa->spa_leaf_list));
+
+	starting_leaf = leaf = list_remove_head(&spa->spa_leaf_list);
+	list_insert_tail(&spa->spa_leaf_list, leaf);
```

Comment:
This LGTM, i only have one question: doesn't `&spa->spa_leaf_list` need the `spa_namespace_lock` held here? My concern is a racing `vdev_remove_child($just_removed_from_list_leaf)` could cause some trouble.

Reply:
`mmp_next_leaf()` is called while `SCL_STATE` is held as `RW_READER`, see module/zfs/mmp.c:366 in `mmp_write_uberblock()`.

That should be sufficient according to the comments at the top of spa_misc.c,  It's also consistent with some of the code paths I examined.  An example is:
`spa_vdev_detach()` -> `vdev_remove_child()`, in which `spa_namespace_lock` and `SCL_ALL` as `RW_WRITER` are held.

I can assert that `SCL_STATE` is held at the top of `mmp_next_leaf()` to ensure future changes don't cause mmp_next_leaf() to run without the lock, and to document the requirement.

Reply:
Hmm... I added that `ASSERT` and it is failing.  Looks like I'm missing something.  Stay tuned...

Reply:
> Hmm... I added that ASSERT and it is failing. Looks like I'm missing something. 

I'd made a typo.  I refreshed the patch.  The only change was adding that ASSERT.

Reply:
Thank you for the clear explanation, LGTM.

Reply:
Thank you for the review.

Reply:
Actually, an `RW_READER` lock here is no longer sufficient since the code is now rotating the leaves.  Since taking a `RW_WRITER` lock is going to be too heavy handed, I think we need to revert to storing a next pointer.  Sorry about that!

Reply:
Darn, you're right.  I'll go back.